overlord/state/warning.go

```diff
@@ -0,0 +1,246 @@
+	DefaultDeleteAfter = time.Hour * 24 * 28
+)
+
+type jsonWarning struct {
+	Message     string     `json:"message"`
+	FirstSeen   time.Time  `json:"first-seen"`
+	LastSeen    time.Time  `json:"last-seen"`
+	LastShown   *time.Time `json:"last-shown,omitempty"`
+	DeleteAfter string     `json:"delete-after,omitempty"`
+	RepeatAfter string     `json:"repeat-after,omitempty"`
+}
+
+type Warning struct {
+	// the warning text itself. Only one of these in the system at a time.
+	message string
+	// the first time one of these messages was created
+	firstSeen time.Time
+	// the last time one of these was created
+	lastSeen time.Time
+	// the last time one of these was shown to the user
+	lastShown time.Time
+	// how much time since one of these was last seen should we drop the message
+	deleteAfter time.Duration
+	// how much time since one of these was last shown should we repeat it
+	repeatAfter time.Duration
+}
+
+func (w *Warning) String() string {
+	return w.message
+}
+
+func (w *Warning) MarshalJSON() ([]byte, error) {
+	jw := jsonWarning{
+		Message:     w.message,
+		FirstSeen:   w.firstSeen,
+		LastSeen:    w.lastSeen,
+		DeleteAfter: w.deleteAfter.String(),
+		RepeatAfter: w.repeatAfter.String(),
+	}
+	if !w.lastShown.IsZero() {
+		jw.LastShown = &w.lastShown
+	}
+
+	return json.Marshal(jw)
+}
+
+func (w *Warning) UnmarshalJSON(data []byte) error {
+	var jw jsonWarning
+	err := json.Unmarshal(data, &jw)
+	if err != nil {
+		return err
+	}
+	w.message = jw.Message
+	w.firstSeen = jw.FirstSeen
+	w.lastSeen = jw.LastSeen
+	if jw.LastShown != nil {
+		w.lastShown = *jw.LastShown
+	}
+	w.deleteAfter, err = time.ParseDuration(jw.DeleteAfter)
+	if err != nil {
+		return err
+	}
+	w.repeatAfter, err = time.ParseDuration(jw.RepeatAfter)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *Warning) IsDeletable(now time.Time) bool {
+	return w.lastSeen.Add(w.deleteAfter).Before(now)
+}
+
+func (w *Warning) IsShowable(t time.Time) bool {
+	return (w.lastShown.IsZero() || w.lastShown.Add(w.repeatAfter).Before(t)) && !w.firstSeen.After(t)
+}
+
+// flattenWarning loops over the warnings map, and returns all
+// warnings therein as a flat list, for serialising.
+// Call with the lock held.
+func (s *State) flattenWarnings() []*Warning {
+	flat := make([]*Warning, 0, len(s.warnings))
+	for _, w := range s.warnings {
+		flat = append(flat, w)
+	}
+	return flat
+}
+
+// unflattenWarnings takes a flat list of warnings and replaces the
+// warning map with them.
+// Call with the lock held.
+func (s *State) unflattenWarnings(flat []*Warning) {
+	s.warnings = make(map[string]*Warning, len(flat))
+	for _, w := range flat {
+		s.warnings[w.message] = w
+	}
+}
+
+// AddWarning records a warning: if it's the first Warning with this
+// message it'll be added (with its firstSeen and lastSeen set to the
+// current time), otherwise the existing one will have its lastSeen
+// updated.
+func (s *State) AddWarning(message string) {
```

Comment:
This will be fun when we start i18n messages. Maybe we should introduce a message id of some kind?

Reply:
the message is the kind is the message (and it'd get l10n'ed on the way out, when we do that)